### PR TITLE
[C++] Add delete operation

### LINKED ISSUE #170: [C++] Support for deletes (port from C#)
We have a tombstone bit in the record header for the C++ version of FASTER, but this needs to be hooked in to a Delete interface to FASTER. This is already implemented in C#, and needs to be ported over in a similar manner.

You can look at the C# version of Deletes at this PR: https://github.com/microsoft/FASTER/pull/114 for the source of the port.

### PR DESCRIPTION
**Pull request by 1u0**
Closes #170.
This is based on #186 PR.

### CODE REVIEW

**Comment by badrishc** [comment]
Thanks for the contribution! We will review shortly :).

File: cc/src/core/internal_contexts.h [hunk 1]
```diff
@@ -326,6 +326,65 @@ class PendingRmwContext : public AsyncPendingRmwContext<typename MC::key_t> {
...
+/// FASTER's internal Delete() context.
+
+/// An internal Delete() context that has gone async and lost its type information.
+template <class K>
+class AsyncPendingDeleteContext : public PendingContext<K> {
+ public:
+  typedef K key_t;
+ protected:
+  AsyncPendingDeleteContext(IAsyncContext& caller_context_, AsyncCallback caller_callback_)
+    : PendingContext<key_t>(OperationType::Read, caller_context_, caller_callback_) {
```
Comment (matthewbrookes) [inline cc/src/core/internal_contexts.h]: Should this be OperationType::Delete?

**Comment by matthewbrookes** [comment]
@1u0 can you merge in master and ping me to re-approve

**Comment by 1u0** [comment]
Thanks for review @matthewbrookes.
I've rebased the branch relative to `master` branch (instead of merging in), sorry for inconvenience.

**Comment by badrishc** [comment]
See this comment. Either of you seen this @1u0 and @matthewbrookes.

Wondering if it got introduced after this and/or the previous PR (clean up of tests).

https://github.com/microsoft/FASTER/pull/193#issuecomment-552205375

**Comment by 1u0** [comment]
I'll take a look at that issue.

**Pull request merged**